### ISSUE TITLE
USDZExporter only exports either channel factor or texture in materials

### ISSUE THREAD

**Issue by optimus007** [mrdoob/three.js#21928]
TL:DR : when a material uses any texture with 0 factor, the usdz exporter ignores the factor and just exports the image . 
example: fully white metalness map with factor 0 looks metallic in usdz

So blender has this quirk where if you only have a roughness texture and no metalness texture(and metallness value is 0) & then you export gltf .
The resulting gltf will have the merged metallicRoughnessAO image with Red channel as white (AO), green as the input roughness texture and blue channel as white (metalness) **with the metallness factor of 0**.
This looks fine as a glb and in three js.
but usdz exporter is either exporting the factor value or the image.

so as temporary fix ,  just like how the color multiplication issue was solved where the resulting base color image is multiplied with the color value we could multiply the factor value with the generated metal,roughness images

**Comment by mrdoob** [comment]
What do you mean with 0 factor?

**Comment by optimus007** [comment]
```
metalness = 0 
metalnessMap=texture
```

model should look fully NON metal..
but in usdz it looks mettalic

**Comment by maciekglowka** [comment]
Hi,
is this blender / metallic issue possibly causing models look too dark in the quick look AR ?
I am having a huge problem with dark models on iOS. I guess it could be coming from unwanted metalness reflectivity x estimated light.

**Comment by optimus007** [comment]
> 
> 
> Hi,
> is this blender / metallic issue possibly causing models look too dark in the quick look AR ?
> I am having a huge problem with dark models on iOS. I guess it could be coming from unwanted metalness reflectivity x estimated light.

try this :In blender put unique or same images in the roughness and metalness channel and then export glb/gltf and make three js usdz export the same ... if it looks fine now, then yes this is the issue i'm referring to 


if there is no metalness texture blender export white image in blue channel and sets the metalness factor as zero .
but in usdz exporter
https://github.com/mrdoob/three.js/blob/ea3762dadd58d239486fcb78a1b3d33252a2e8cf/examples/jsm/exporters/USDZExporter.js#L440-L450

only the image OR  factor is exported  ...these models just end exporting the white image thus leading to metallic looking usdzs

**Comment by maciekglowka** [comment]
Thanks a lot,
indeed when I crank the metalness in Blender up to 1 the model starts to look as if it had the same problems as the one in Quick Look.
I am using usally a green channel only roughness map (so blue, red black) as recommended in the GLTF exporter docs. So I can try to connect the blue channel in blender to the metallic input. Will do some testing tomorrow.

**Comment by mrdoob** [comment]
I think #22201 should solve this.

**Comment by maciekglowka** [comment]
Thanks.
For now I have solved by connecting blue output of the texture into metalness - whenever I have a green channel roughness map.
So definitely the problem was metalness being set to 1.

**Comment by mrdoob** [comment]
Cool! Closing this one then.

@optimus007 please confirm that it solves the issue if you can 👍

**Comment by optimus007** [comment]
@mrdoob 

if roughness value === 1 & texture is present image is exported otherwise only the value is exported correct ?
as a temp fix this is fine  

but if the editor  wants roughness map and roughness = 0.5 ,then usdz wont have the image, leading to incorrect model again .

Another solution could be:
Just like how in the diffuse texture the material color is passed in buildTexture function , a factor value could also be passed which is multiplied with all pixel values.
https://github.com/mrdoob/three.js/blob/e203359dc86bb1884d6cb49bddabc3bc6c6fb973/examples/jsm/exporters/USDZExporter.js#L392

so something like :

`function buildTexture( texture, mapType, color,factor ){...}`

and then 

roughnessMap=texture & roughness=0.8 will do `buildTexture (material.roughnessMap, 'roughness', null, material.roughness )`
or 
metalnessMap=texture & metalness =0 will do  `buildTexture ( material.metalnessMap, 'metallic', null ,material.metalness )` resulting in a black image (image equivalent of metallness =0 )

**Comment by mrdoob** [comment]
It's a bit more complicated than that because the same texture can be reused by `roughnessMap` and `metalnessMap`, so the code needs to modify only the relevant color channels.

I think I'll leave this closed until more people bumps into this.

**Comment by optimus007** [comment]
True, i though it would be a simple fix ,leave it for now then.

for anyone else testing this issue in the future 
Here's a glb file . Export it as usdz via THREE usdz exporter and you'll notice 3rd and 4th cube does not look the same as this image .
That's the problem highlighted in this issue  & might be the reason your usdz looks dark

![glb](https://user-images.githubusercontent.com/6885294/127433830-82b397b7-3abf-465d-b742-c53e9b96f50c.png)

[test.zip](https://github.com/mrdoob/three.js/files/6897974/test.zip)
